# **Ejercicios de pair programming Módulo 3 Sprint 2**
## **Tableau**
### **Limpieza del dataframe**

---

In [1]:
import pandas as pd
import numpy as nd
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
cuernos = pd.read_csv("../datos/affairs.csv", index_col = 0)

In [3]:
def df_explore(dataframe):
    print(
        f"El dataframe tiene {dataframe.shape[0]} filas y {dataframe.shape[1]} columnas")
    print("--------------")
    print(f"El dataframe tiene las siguientes columnas:")
    for col in dataframe.columns:
        print(col)
    print("--------------")
    print(f"Los principales estadísticos de las variables numéricas son:")
    display(dataframe.describe().T)
    print("--------------")
    print(f"Los principales estadísticos de las variables categóricas son:")
    display(dataframe.describe(include='object').T)
    print("--------------")
    print("El porcentaje de nulos por columna:")
    for i, col in enumerate(dataframe.isnull().sum()):
        print(
            f"{dataframe.isnull().sum().index[i]}: {col/dataframe.shape[0]*100}")
    print("--------------")
    print(f"Hay {dataframe.duplicated().sum()} filas duplicadas.")
    print("--------------")
    print("A continuación el resultado del método .info():")
    dataframe.info()

In [4]:
df_explore(cuernos)

El dataframe tiene 601 filas y 9 columnas
--------------
El dataframe tiene las siguientes columnas:
affairs
gender
age
yearsmarried
children
religiousness
education
occupation
rating
--------------
Los principales estadísticos de las variables numéricas son:


count       mean       std     min   25%   50%   75%   max
affairs        601.0   1.455907  3.298758   0.000   0.0   0.0   0.0  12.0
age            601.0  32.487521  9.288762  17.500  27.0  32.0  37.0  57.0
yearsmarried   601.0   8.177696  5.571303   0.125   4.0   7.0  15.0  15.0
religiousness  601.0   3.116473  1.167509   1.000   2.0   3.0   4.0   5.0
education      601.0  16.166389  2.402555   9.000  14.0  16.0  18.0  20.0
occupation     601.0   4.194676  1.819443   1.000   3.0   5.0   6.0   7.0
rating         601.0   3.931780  1.103179   1.000   3.0   4.0   5.0   5.0

--------------
Los principales estadísticos de las variables categóricas son:


count unique     top freq
gender     601      2  female  315
children   601      2     yes  430

--------------
El porcentaje de nulos por columna:
affairs: 0.0
gender: 0.0
age: 0.0
yearsmarried: 0.0
children: 0.0
religiousness: 0.0
education: 0.0
occupation: 0.0
rating: 0.0
--------------
Hay 24 filas duplicadas.
--------------
A continuación el resultado del método .info():
<class 'pandas.core.frame.DataFrame'>
Int64Index: 601 entries, 4 to 9010
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   affairs        601 non-null    int64  
 1   gender         601 non-null    object 
 2   age            601 non-null    float64
 3   yearsmarried   601 non-null    float64
 4   children       601 non-null    object 
 5   religiousness  601 non-null    int64  
 6   education      601 non-null    int64  
 7   occupation     601 non-null    int64  
 8   rating         601 non-null    int64  
dtypes: float64(2), int64(5), object(2)
memory usage: 47.0+ KB


In [5]:
cuernos.head()

affairs  gender   age  yearsmarried children  religiousness  education  \
4         0    male  37.0         10.00       no              3         18   
5         0  female  27.0          4.00       no              4         14   
11        0  female  32.0         15.00      yes              1         12   
16        0    male  57.0         15.00      yes              5         18   
23        0    male  22.0          0.75       no              2         17   

    occupation  rating  
4            7       4  
5            6       4  
11           1       4  
16           6       5  
23           6       3

In [6]:
cuernos.reset_index(drop = True, inplace=True)

**Description**

Infidelity data, known as Fair's Affairs. Cross-section data from a survey conducted by Psychology Today in 1969.

**Format**

A data frame containing 601 observations on 9 variables.

**affairs**

numeric. How often engaged in extramarital sexual intercourse during the past year? 0 = none, 1 = once, 2 = twice, 3 = 3 times, 7 = 4–10 times, 12 = monthly, 12 = weekly, 12 = daily.

**gender**

factor indicating gender.

**age**

numeric variable coding age in years: 17.5 = under 20, 22 = 20–24, 27 = 25–29, 32 = 30–34, 37 = 35–39, 42 = 40–44, 47 = 45–49, 52 = 50–54, 57 = 55 or over.

**yearsmarried**

numeric variable coding number of years married: 0.125 = 3 months or less, 0.417 = 4–6 months, 0.75 = 6 months–1 year, 1.5 = 1–2 years, 4 = 3–5 years, 7 = 6–8 years, 10 = 9–11 years, 15 = 12 or more years.

**children**

factor. Are there children in the marriage?

**religiousness**

numeric variable coding religiousness: 1 = anti, 2 = not at all, 3 = slightly, 4 = somewhat, 5 = very.

**education**

numeric variable coding level of education: 9 = grade school, 12 = high school graduate, 14 = some college, 16 = college graduate, 17 = some graduate work, 18 = master's degree, 20 = Ph.D., M.D., or other advanced degree.

**occupation**

numeric variable coding occupation according to Hollingshead classification (reverse numbering).

**rating**

numeric variable coding self rating of marriage: 1 = very unhappy, 2 = somewhat unhappy, 3 = average, 4 = happier than average, 5 = very happy.

In [7]:
diccionario_age = {17.5 : 'Menor de 20', 22 : '20–24', 27 : '25–29', 32 : '30–34', 37 : '35–39', 42 : '40–44', 47 : '45–49', 52 : '50–54', 57 : '55 o más'}

In [8]:
diccionario_affaris = {0 : 'Nunca', 1 : '1 vez', 2 : '2 veces', 3 : '3 veces', 7 : '4–10 veces', 12 : 'Mensualmente', 12 : 'Semanalmente', 12 : 'Diariamente'}

diccionario_yearsmarried = {0.125 : '3 meses o menos', 0.417 : '4–6 meses', 0.75 : '6 meses–1 año', 1.5 : '1–2 años', 4 : '3–5 años', 7 : '6–8 años', 10 : '9–11 años', 15 : '12 o más años'}

diccionario_religiousness = {1 : 'Anti religioso', 2 : 'Nada religioso', 3 : 'Un poco religioso', 4 : 'Bastante religioso', 5 : 'Muy religioso'}

diccionario_education = {9 : 'Estudios primarios', 12 : 'Estudios secundarios', 14 : 'Alguna educación superior', 16 : 'Estudios universitarios', 17 : 'Algún estudio de posgrado', 18 : "Maestría", 20 : "Doctorado"}

diccionario_rating = {1 : 'Muy infeliz', 2 : 'Un poco infeliz', 3 : 'Felicidad media', 4 : 'Más feliz que la media', 5 : 'Muy feliz'}

diccionario_occupation = {1: 'Altos ejecutivos', 2: 'Jefes de negocio', 3: 'Personal administrativo', 4: 'Clero, ventas y técnicos', 5: 'Trabajadores manuales capacitados', 6:'Operarios semicapacitados', 7:'Trabajadores no cualificados'}

diccionario_children = {'no':'No', 'yes':'Sí'}

diccionario_gender = {'male':'Hombre', 'female':'Mujer'}

In [9]:
cuernos.head()

affairs  gender   age  yearsmarried children  religiousness  education  \
0        0    male  37.0         10.00       no              3         18   
1        0  female  27.0          4.00       no              4         14   
2        0  female  32.0         15.00      yes              1         12   
3        0    male  57.0         15.00      yes              5         18   
4        0    male  22.0          0.75       no              2         17   

   occupation  rating  
0           7       4  
1           6       4  
2           1       4  
3           6       5  
4           6       3

In [10]:
lista_dicc = [diccionario_affaris, diccionario_gender, diccionario_yearsmarried, diccionario_children, diccionario_religiousness, diccionario_education, diccionario_occupation, diccionario_rating]

df_age = cuernos.drop('age', axis=1)

for i, col in enumerate(df_age):
    df_age[col] = df_age[col].map(lista_dicc[i])

In [11]:
df_age.head()

affairs  gender   yearsmarried children       religiousness  \
0   Nunca  Hombre      9–11 años       No   Un poco religioso   
1   Nunca   Mujer       3–5 años       No  Bastante religioso   
2   Nunca   Mujer  12 o más años       Sí      Anti religioso   
3   Nunca  Hombre  12 o más años       Sí       Muy religioso   
4   Nunca  Hombre  6 meses–1 año       No      Nada religioso   

                   education                    occupation  \
0                   Maestría  Trabajadores no cualificados   
1  Alguna educación superior     Operarios semicapacitados   
2       Estudios secundarios              Altos ejecutivos   
3                   Maestría     Operarios semicapacitados   
4  Algún estudio de posgrado     Operarios semicapacitados   

                   rating  
0  Más feliz que la media  
1  Más feliz que la media  
2  Más feliz que la media  
3               Muy feliz  
4         Felicidad media

In [12]:
cuernos['age_cat'] = cuernos['age'].map(diccionario_age)

In [13]:
cuernos[df_age.columns] = df_age

In [14]:
cuernos.head()

affairs  gender   age   yearsmarried children       religiousness  \
0   Nunca  Hombre  37.0      9–11 años       No   Un poco religioso   
1   Nunca   Mujer  27.0       3–5 años       No  Bastante religioso   
2   Nunca   Mujer  32.0  12 o más años       Sí      Anti religioso   
3   Nunca  Hombre  57.0  12 o más años       Sí       Muy religioso   
4   Nunca  Hombre  22.0  6 meses–1 año       No      Nada religioso   

                   education                    occupation  \
0                   Maestría  Trabajadores no cualificados   
1  Alguna educación superior     Operarios semicapacitados   
2       Estudios secundarios              Altos ejecutivos   
3                   Maestría     Operarios semicapacitados   
4  Algún estudio de posgrado     Operarios semicapacitados   

                   rating   age_cat  
0  Más feliz que la media     35–39  
1  Más feliz que la media     25–29  
2  Más feliz que la media     30–34  
3               Muy feliz  55 o más  
4         Felicidad media     20–24

In [15]:
cuernos['age'].unique()

array([37. , 27. , 32. , 57. , 22. , 47. , 42. , 52. , 17.5])

In [16]:
break

SyntaxError: 'break' outside loop (668683560.py, line 1)

In [17]:
cuernos['age'] = cuernos['age'].map({37.:37 , 27.:27 , 32.:32 , 57.:57 , 22.:22 , 47.:47 , 42.:42 , 52.:52 , 17.5:18})

In [18]:
cuernos['age'].unique()

array([37, 27, 32, 57, 22, 47, 42, 52, 18])

In [19]:
cuernos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 601 entries, 0 to 600
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   affairs        601 non-null    object
 1   gender         601 non-null    object
 2   age            601 non-null    int64 
 3   yearsmarried   601 non-null    object
 4   children       601 non-null    object
 5   religiousness  601 non-null    object
 6   education      601 non-null    object
 7   occupation     601 non-null    object
 8   rating         601 non-null    object
 9   age_cat        601 non-null    object
dtypes: int64(1), object(9)
memory usage: 47.1+ KB


In [20]:
cuernos.to_csv('../datos/affairs_procesado.csv')